ledger/txtail.go

```diff
@@ -227,29 +260,50 @@ func (t *txTail) commitRound(ctx context.Context, tx *sql.Tx, dcc *deferredCommi
 
 	t.tailMu.RLock()
 	for i := uint64(0); i < dcc.offset; i++ {
-		roundsData = append(roundsData, t.roundTailSerializedData[i])
+		roundsData = append(roundsData, t.roundTailSerializedDeltas[i])
 	}
 	// get the MaxTxnLife from the consensus params of the latest round in this commit range
-	maxTxnLifeRound := basics.Round(config.Consensus[t.consensusVersions[dcc.offset]].MaxTxnLife) + 1
+	// preserve data for MaxTxnLife + DeeperBlockHeaderHistory
+	rnd := dcc.oldBase + basics.Round(dcc.offset)
+	proto, ok := config.Consensus[t.blockHeaderData[rnd].CurrentProtocol]
+	if !ok {
+		return fmt.Errorf("round %d not found in blockHeaderData: lowest=%d, base=%d", rnd, t.lowestBlockHeaderRound, dcc.oldBase)
+	}
+	retainSize := proto.MaxTxnLife + proto.DeeperBlockHeaderHistory
 	t.tailMu.RUnlock()
 
-	forgetRound := (baseRound + basics.Round(dcc.offset)).SubSaturate(maxTxnLifeRound)
-	if err := txtailNewRound(ctx, tx, baseRound, roundsData, forgetRound); err != nil {
+	// determine the round to remove data
+	// This is the similar formula to the committedUpTo: rnd + 1 - retain size
+	forgetBeforeRound := (baseRound + basics.Round(dcc.offset)).SubSaturate(basics.Round(retainSize))
+	if err := txtailNewRound(ctx, tx, baseRound, roundsData, forgetBeforeRound); err != nil {
 		return fmt.Errorf("txTail: unable to persist new round %d : %w", baseRound, err)
 	}
 	return nil
 }
 
 func (t *txTail) postCommit(ctx context.Context, dcc *deferredCommitContext) {
 	t.tailMu.Lock()
-	t.roundTailSerializedData = t.roundTailSerializedData[dcc.offset:]
+	rnd := dcc.oldBase + basics.Round(dcc.offset)
+
+	t.roundTailSerializedDeltas = t.roundTailSerializedDeltas[dcc.offset:]
+
 	// get the MaxTxnLife from the consensus params of the latest round in this commit range
-	maxTxnLife := config.Consensus[t.consensusVersions[dcc.offset]].MaxTxnLife
-	// keep the latest 1001 entries on the hash, before roundTailSerializedData
-	firstTailIdx := len(t.roundTailHashes) - len(t.roundTailSerializedData) - int(maxTxnLife)
-	if firstTailIdx > 0 {
-		t.roundTailHashes = t.roundTailHashes[firstTailIdx:]
-		t.consensusVersions = t.consensusVersions[firstTailIdx:]
+	// preserve data for MaxTxnLife + DeeperBlockHeaderHistory
+	proto := config.Consensus[t.blockHeaderData[rnd].CurrentProtocol]
+	retainSize := proto.MaxTxnLife + proto.DeeperBlockHeaderHistory
+	newLowestRound := rnd.SubSaturate(basics.Round(retainSize)) + 1
```

Comment:
```suggestion
	newLowestRound := rnd.SubSaturate(basics.Round(retainSize) + 1)
```

Reply:
no. say newBase = 10 and retain size is 3. Then the map should have 8, 9, 10 and newLowestRound = 8.
so newLowestRound = 10 - 3 + 1 = 8.

Reply:
Sorry, I meant `newLowestRound := rnd.SubSaturate(basics.Round(retainSize) - 1)`. The reason is that we don't want to remove round 0 when we shouldn't.

Alternatively:
```
newLowestRound := 0
if rnd + 1 > retainSize {
  newLowestRound = rnd + 1 - retainSize
}
```

Reply:
this makes sense, thank you for noticing.